File: OpenGlStudy/粒子特效/Mist_雾/OpenGLEngine/Model/Mountion.cpp
//
//  Mountion.cpp
//  Mist
//
//  Created by 苏莫离 on 2019/5/15.
//

#include "Mountion.hpp"
#include <math.h>
#include "ShaderTool.h"
#include "GLMatrixState.hpp"
#import <simd/simd.h>
using namespace std;

float Mountion::toRadians(float angle) { //将角度转换为弧度
    return angle * 3.1415926f / 180;
}

Mountion::Mountion(vector<vector<float>> yArray) {
    initVertexData(yArray); // 初始化顶点数据
    initShader(); // 初始化着色器
}

/// 矩形：由大量的小三角形组成
void Mountion::initVertexData(vector<vector<float>> yArray) {
    unsigned long rows = yArray.size() - 1, cols = yArray[0].size() - 1;
    vCount = (int)(rows * cols * 6);
    float sizeWidth = 16.0f / rows, sizeHeight = 16.0f / cols;
    
    vector_float3 *vertices = new vector_float3[vCount];
    vector_float2 *textureCoors = new vector_float2[vCount];

    int verticeIndex = 0;//顶点计数器
    for(int j = 0; j < rows; j++) {
        for(int i = 0; i < cols; i++) {
            //计算当前格子左上侧点坐标
            float zsx = -unitSize * cols / 2 + i * unitSize;
            float zsz = -unitSize * rows / 2 + j * unitSize;
            
            float s = j * sizeWidth, t = i * sizeHeight;
            
            textureCoors[verticeIndex] = (vector_float2){s, t};
            vertices[verticeIndex++] = (vector_float3){zsx, yArray[j][i], zsz};
            
            textureCoors[verticeIndex] = (vector_float2){s, t + sizeHeight};
            vertices[verticeIndex++] = (vector_float3){zsx, yArray[j+1][i], zsz + unitSize};
            
            textureCoors[verticeIndex] = (vector_float2){s + sizeWidth, t};
            vertices[verticeIndex++] = (vector_float3){zsx+unitSize,yArray[j][i+1], zsz};

            textureCoors[verticeIndex] = (vector_float2){s + sizeWidth, t};
            vertices[verticeIndex++] = (vector_float3){zsx+unitSize,yArray[j][i+1], zsz};
            
            textureCoors[verticeIndex] = (vector_float2){s, t + sizeHeight};
            vertices[verticeIndex++] = (vector_float3){zsx,yArray[j+1][i], zsz+unitSize};
            
            textureCoors[verticeIndex] = (vector_float2){s + sizeWidth, t + sizeHeight};
            vertices[verticeIndex++] = (vector_float3){zsx+unitSize,yArray[j+1][i+1], zsz+unitSize};
        }
    }

    mVertexBuffer = vertices;
    mTexCoorBuffer = textureCoors;
}

void Mountion::initShader() {
    //基于顶点着色器与片元着色器创建渲染管线
    mProgram = ShaderTool::createProgram("VertexShader", "FragmentShader");
    maPositionHandle = glGetAttribLocation(mProgram, "aPosition");
    maTexCoorHandle = glGetAttribLocation(mProgram, "aTexCoor");
    muMVPMatrixHandle = glGetUniformLocation(mProgram, "uMVPMatrix");
    muMMatrixHandle = glGetUniformLocation(mProgram, "uMMatrix");
    //获取程序中摄像机位置引用的id
    muCamaraLocationHandle = glGetUniformLocation(mProgram, "uCamaraLocation");
    //获取程序中体积雾产生者平面高度引用的id
    slabYHandle = glGetUniformLocation(mProgram, "slabY");
    //获取程序中体积雾高度扰动起始角引用的id
    startAngleHandle = glGetUniformLocation(mProgram, "startAngle");
    
    //纹理
    landStartYYHandle = glGetUniformLocation(mProgram, "landStartY"); //x位置
    landYSpanHandle = glGetUniformLocation(mProgram, "landYSpan");  //x最大
}

void Mountion::drawSelf(void) {
    glUseProgram(mProgram);//指定使用某套着色器程序
    
    glUniformMatrix4fv(muMVPMatrixHandle, 1, GL_FALSE, GLMatrixState::getFinalMatrix()); //将最终变换矩阵传入shader程序
    glUniformMatrix4fv(muMMatrixHandle, 1, GL_FALSE, GLMatrixState::getMMatrix()); //将位置、旋转变换矩阵传入着色器程序
    glUniform3fv(muCamaraLocationHandle, 1, GLMatrixState::cameraFB); //将摄像机位置传入着色器程序
    
    /** 首先将体积雾所需雾平面的高度传入渲染管线，
     * 接着将用于扰动雾平面高度的正弦曲线所需起始角传入渲染管线。
     * 最后增加起始角的值，并通过取模的方式将起始角限制在 0~360 的范围内
     */
    glUniform1f(slabYHandle, TJ_GOG_SLAB_Y); //将体积雾的雾平面高度传入渲染管线
    glUniform1f(startAngleHandle, toRadians(startAngle));  //将体积雾扰动起始角传入渲染管线
    startAngle = startAngle + 90.0f;
    if (startAngle > 360.0f ) {
        startAngle -= 360.0f;
    }
    glVertexAttribPointer(maPositionHandle,3, GL_FLOAT, GL_FALSE,3 * sizeof(float), mVertexBuffer); /// 指定顶点数据
    glVertexAttribPointer(maTexCoorHandle,2, GL_FLOAT, GL_FALSE,2 * sizeof(float), mTexCoorBuffer); /// 指定纹理数据
    glEnableVertexAttribArray(maPositionHandle); //启用顶点位置数据数组
    glEnableVertexAttribArray(maTexCoorHandle);
    
    //传送相应的x参数
    glUniform1f(landStartYYHandle, 0);
    glUniform1f(landYSpanHandle, 50);
    
    //绘制纹理矩形
    glDrawArrays(GL_TRIANGLES, 0, vCount);
}

File: OpenGlStudy/下册源码/第03章 片元着色器的妙用/Sample3_13/src/com/bn/Sample3_13/Mountion.java
package com.bn.Sample3_13;

import java.nio.ByteBuffer;
import java.nio.ByteOrder;
import java.nio.FloatBuffer;
import android.opengl.GLES30;

public class Mountion
{
	//单位长度
	float UNIT_SIZE=3.0f;
	//体积雾产生者平面的高度
	float TJ_GOG_SLAB_Y=8f;
	//扰动起始角
	float startAngle=0;

	//自定义渲染管线的id
	int mProgram;
	//总变化矩阵引用的id
	int muMVPMatrixHandle;
	//基本变换阵引用的id
	int muMMatrixHandle;
	//摄像机位置引用的id
	int muCamaraLocationHandle;
	//体积雾产生者平面高度引用的id
	int slabYHandle;
	//体积雾高度扰动起始角引用的id
	int startAngleHandle;
	//顶点位置属性引用id
	int maPositionHandle;
	//顶点纹理坐标属性引用id
	int maTexCoorHandle;

	//草地的id
	int sTextureGrassHandle;
	//石头的id
	int sTextureRockHandle;
	//起始x值
	int landStartYYHandle;
	//长度
	int landYSpanHandle;

	//顶点数据缓冲和纹理坐标数据缓冲
	FloatBuffer mVertexBuffer;
	FloatBuffer mTexCoorBuffer;
	//顶点数量
	int vCount=0;

	public Mountion(MySurfaceView mv,float[][] yArray,int rows,int cols)
	{
		initVertexData(yArray,rows,cols);
		initShader(mv);
	}
	//初始化顶点坐标与着色数据的方法
    public void initVertexData(float[][] yArray,int rows,int cols)
    {
    	//顶点坐标数据的初始化
    	vCount=cols*rows*2*3;//每个格子两个三角形，每个三角形3个顶点
        float vertices[]=new float[vCount*3];//每个顶点xyz三个坐标
        int count=0;//顶点计数器
        for(int j=0;j<rows;j++)
        {
        	for(int i=0;i<cols;i++)
        	{
        		//计算当前格子左上侧点坐标
        		float zsx=-UNIT_SIZE*cols/2+i*UNIT_SIZE;
        		float zsz=-UNIT_SIZE*rows/2+j*UNIT_SIZE;

        		vertices[count++]=zsx;
        		vertices[count++]=yArray[j][i];
        		vertices[count++]=zsz;

        		vertices[count++]=zsx;
        		vertices[count++]=yArray[j+1][i];
        		vertices[count++]=zsz+UNIT_SIZE;

        		vertices[count++]=zsx+UNIT_SIZE;
        		vertices[count++]=yArray[j][i+1];
        		vertices[count++]=zsz;

        		vertices[count++]=zsx+UNIT_SIZE;
        		vertices[count++]=yArray[j][i+1];
        		vertices[count++]=zsz;

        		vertices[count++]=zsx;
        		vertices[count++]=yArray[j+1][i];
        		vertices[count++]=zsz+UNIT_SIZE;

        		vertices[count++]=zsx+UNIT_SIZE;
        		vertices[count++]=yArray[j+1][i+1];
        		vertices[count++]=zsz+UNIT_SIZE;
        	}
        }

        //创建顶点坐标数据缓冲
        ByteBuffer vbb = ByteBuffer.allocateDirect(vertices.length*4);
        vbb.order(ByteOrder.nativeOrder());//设置字节顺序
        mVertexBuffer = vbb.asFloatBuffer();//转换为Float型缓冲
        mVertexBuffer.put(vertices);//向缓冲区中放入顶点坐标数据
        mVertexBuffer.position(0);//设置缓冲区起始位置

        //顶点纹理坐标数据的初始化
        float[] texCoor=generateTexCoor(cols,rows);
        //创建顶点纹理坐标数据缓冲
        ByteBuffer cbb = ByteBuffer.allocateDirect(texCoor.length*4);
        cbb.order(ByteOrder.nativeOrder());//设置字节顺序
        mTexCoorBuffer = cbb.asFloatBuffer();//转换为Float型缓冲
        mTexCoorBuffer.put(texCoor);//向缓冲区中放入顶点着色数据
        mTexCoorBuffer.position(0);//设置缓冲区起始位置
    }

	//初始化Shader的方法
	public void initShader(MySurfaceView mv)
	{
		String mVertexShader=ShaderUtil.loadFromAssetsFile("vertex.sh", mv.getResources());
		String mFragmentShader=ShaderUtil.loadFromAssetsFile("frag.sh", mv.getResources());
		//基于顶点着色器与片元着色器创建程序
        mProgram = ShaderUtil.createProgram(mVertexShader, mFragmentShader);
        //获取程序中顶点位置属性引用id
        maPositionHandle = GLES30.glGetAttribLocation(mProgram, "aPosition");
        //获取程序中顶点纹理坐标属性引用id
        maTexCoorHandle= GLES30.glGetAttribLocation(mProgram, "aTexCoor");
        //获取程序中总变换矩阵引用的id
        muMVPMatrixHandle = GLES30.glGetUniformLocation(mProgram, "uMVPMatrix");
        //获取程序中基本变换矩阵引用的id
        muMMatrixHandle=GLES30.glGetUniformLocation(mProgram, "uMMatrix");
        //获取程序中摄像机位置引用的id
        muCamaraLocationHandle=GLES30.glGetUniformLocation(mProgram, "uCamaraLocation");
        //获取程序中体积雾产生者平面高度引用的id
        slabYHandle=GLES30.glGetUniformLocation(mProgram, "slabY");
        //获取程序中体积雾高度扰动起始角引用的id
        startAngleHandle=GLES30.glGetUniformLocation(mProgram, "startAngle");

        //纹理
		//草地
		sTextureGrassHandle=GLES30.glGetUniformLocation(mProgram, "sTextureGrass");
		//石头
		sTextureRockHandle=GLES30.glGetUniformLocation(mProgram, "sTextureRock");
		//x位置
		landStartYYHandle=GLES30.glGetUniformLocation(mProgram, "landStartY");
		//x最大
		landYSpanHandle=GLES30.glGetUniformLocation(mProgram, "landYSpan");
	}

	//自定义的绘制方法drawSelf
	public void drawSelf(int texId,int rock_textId)
	{
		//指定使用某套着色器程序
   	 	GLES30.glUseProgram(mProgram);
        //将最终变换矩阵传入渲染管线
        GLES30.glUniformMatrix4fv(muMVPMatrixHandle, 1, false, MatrixState.getFinalMatrix(), 0);
        //将基本变换矩阵传入渲染管线
        GLES30.glUniformMatrix4fv(muMMatrixHandle, 1, false, MatrixState.getMMatrix(), 0);
        //将摄像机位置传入渲染管线
        GLES30.glUniform3fv(muCamaraLocationHandle, 1,MatrixState.cameraFB);


        //将体积雾的雾平面高度传入渲染管线
        GLES30.glUniform1f(slabYHandle, TJ_GOG_SLAB_Y);
        //将体积雾扰动起始角传入渲染管线
        GLES30.glUniform1f(startAngleHandle, (float) Math.toRadians(startAngle));
        //修改扰动角的值，每次加3，取值范围永远在0~360的范围内
        startAngle=(startAngle+3f)%360.0f;

        //传送顶点位置数据
		GLES30.glVertexAttribPointer
		(
			maPositionHandle,
			3,
			GLES30.GL_FLOAT,
			false,
			3*4,
			mVertexBuffer
		);
		//传送顶点纹理坐标数据
		GLES30.glVertexAttribPointer
		(
			maTexCoorHandle,
			2,
			GLES30.GL_FLOAT,
			false,
			2*4,
			mTexCoorBuffer
		);
		//允许顶点位置数据数组
        GLES30.glEnableVertexAttribArray(maPositionHandle);
        GLES30.glEnableVertexAttribArray(maTexCoorHandle);

        //绑定纹理
        GLES30.glActiveTexture(GLES30.GL_TEXTURE0);
        GLES30.glBindTexture(GLES30.GL_TEXTURE_2D, texId);
        GLES30.glActiveTexture(GLES30.GL_TEXTURE1);
		GLES30.glBindTexture(GLES30.GL_TEXTURE_2D, rock_textId);
		GLES30.glUniform1i(sTextureGrassHandle, 0);//使用0号纹理
        GLES30.glUniform1i(sTextureRockHandle, 1); //使用1号纹理

        //传送相应的x参数
        GLES30.glUniform1f(landStartYYHandle, 0);
        GLES30.glUniform1f(landYSpanHandle, 50);

        //绘制纹理矩形
        GLES30.glDrawArrays(GLES30.GL_TRIANGLES, 0, vCount);
	}
	//自动切分纹理产生纹理数组的方法
    public float[] generateTexCoor(int bw,int bh)
    {
    	float[] result=new float[bw*bh*6*2];
    	float sizew=16.0f/bw;//列数
    	float sizeh=16.0f/bh;//行数
    	int c=0;
    	for(int i=0;i<bh;i++)
    	{
    		for(int j=0;j<bw;j++)
    		{
    			//每行列一个矩形，由两个三角形构成，共六个点，12个纹理坐标
    			float s=j*sizew;
    			float t=i*sizeh;

    			result[c++]=s;
    			result[c++]=t;

    			result[c++]=s;
    			result[c++]=t+sizeh;

    			result[c++]=s+sizew;
    			result[c++]=t;

    			result[c++]=s+sizew;
    			result[c++]=t;

    			result[c++]=s;
    			result[c++]=t+sizeh;

    			result[c++]=s+sizew;
    			result[c++]=t+sizeh;
    		}
    	}
    	return result;
    }
}
